### PR TITLE
Implement Pagination

### LINKED ISSUE #18: tasks missing?
My API connectivity is working ok...

```
gkendal@5005331:~$ dovico --myself
Informations about yourself
 - ID:         3240
 - First Name: Geoff
 - Last Name:  Kendal
```

But whenever I'm interacting with the actual projects etc...

```
gkendal@5005331:~$ dovico --show --start 2019-09-16 --end 2019-09-19
== List of Time Entries between 2019-09-16 and 2019-09-19 ==
Traceback (most recent call last):
	9: from /Users/gkendal/.rbenv/versions/2.6.3/bin/dovico:23:in `<main>'
	8: from /Users/gkendal/.rbenv/versions/2.6.3/bin/dovico:23:in `load'
	7: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/bin/dovico:4:in `<top (required)>'
	6: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:77:in `run'
	5: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:123:in `display_time_entries'
	4: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:10:in `format_entries'
	3: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:10:in `map'
	2: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:12:in `block in format_entries'
	1: from /Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:20:in `time_entry_text'
/Users/gkendal/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:36:in `project_task': undefined method `tasks' for nil:NilClass (NoMethodError)
```

I can list projects and tasks ok, but we seem to have an additional level under here which seems to be missing from when I'm trying to interact with them... Any advice?

### LINKED ISSUE #19: Implement pagination for assignments
Currently, we only fetch assignments returned by the API in a single call.

If there are too many assignments (>2000 ?), the results are paginated by the Dovico API, but we never fetch the 2nd page.

We need to implement pagination. Feel free to contribute if you want to!

### PR DESCRIPTION
**Pull request by chtitux**
It should fix #18 and fix #19 

Documentation: https://www.dovico.com/developer/API_doc/#t=API_calls_older%2FProjects_v5.html%23Get%20all%20Projectsbc-1&rhtocid=_3_12_0

### CODE REVIEW

**Comment by geftactics** [comment]
May have some issue?

```
14:56 $ dovico --show --start 2019-09-16 --end 2019-09-19
== List of Time Entries between 2019-09-16 and 2019-09-19 ==
== Error during HTTP request ==
Status:      BadRequest
Description: Please specify the "version" value in the correct numeric format.

Traceback (most recent call last):
	10: from /Users/gkendal/.rbenv/versions/2.6.5/bin/dovico:23:in `<main>'
	 9: from /Users/gkendal/.rbenv/versions/2.6.5/bin/dovico:23:in `load'
	 8: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/bin/dovico:4:in `<top (required)>'
	 7: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:77:in `run'
	 6: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:122:in `display_time_entries'
	 5: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry.rb:43:in `search'
	 4: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/api_client.rb:36:in `get_paginated_list'
	 3: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/api_client.rb:36:in `loop'
	 2: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/api_client.rb:40:in `block in get_paginated_list'
	 1: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/api_client.rb:17:in `get'
/Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/api_client.rb:86:in `perform!': Error during HTTP request (RuntimeError)
```

**Comment by geftactics** [comment]
Did a little more debugging, I see a load of http requests to URLs like:
`https://api.dovico.com/Assignments/A123`
`https://api.dovico.com/Assignments/B456`

Then the final request is to:
`https://api.dovico.com/TimeEntries/?next=M123456&daterange=2019-10-01%202019-10-07&version=5`

...which seems to generate the above error?

**Comment by chtitux** [comment]
@squiggleuk I think the issue was, in the previous implementation, the `version=5` was added twice (once because it's passed in the `NextParamURI`, once because we add it on every request).

I have edited the patch so now, we merge the query parameters : https://github.com/trainline-eu/dovico/pull/20/files#diff-b0c33809fe07c1db9df34b54335e150bR72-R77

Sadly, WebMock parses the URL and deduplicate the query parameters so specs can't catch that error :(

could you test again and tell me if it's working or not?
I have sadly no pagination in my Dovico setup.

File: lib/dovico/api_client.rb [hunk 1]
```diff
@@ -28,6 +30,21 @@ def delete(path, params: {}, body: nil)
         perform!(:delete, path, params: params, body: body)
       end
 
+      def get_paginated_list(path, object_key, **params)
+        results = get(path, **params)
+        objects = results[object_key]
+
+        loop do
+          next_page_uri = results["NextPageURI"]
+          break if next_page_uri.nil? || next_page_uri == "N/A"
```
Comment (misdoro) [inline lib/dovico/api_client.rb:39]: :thinking: Should we check if `next_page_uri` doesn't look like an URL ?

Comment (chtitux) [reply lib/dovico/api_client.rb:39]: Well, the possible values are : 
- an URL
- `N/A`

Maybe some API calls does not include the field.

So I find more easy to just test it's exist and it's not `N/A`. I think we can trust the Dovico API here.

**Comment by geftactics** [comment]
API version error seems to be corrected! Thanks. I see another issue now, but I'm not sure if its related to pagination though... Should I open another issue for this?

```
07:28 $ dovico --show --start 2019-10-01 --end 2019-10-07 --debug
== List of Time Entries between 2019-10-01 and 2019-10-07 ==
Traceback (most recent call last):
	9: from /Users/gkendal/.rbenv/versions/2.6.5/bin/dovico:23:in `<main>'
	8: from /Users/gkendal/.rbenv/versions/2.6.5/bin/dovico:23:in `load'
	7: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/bin/dovico:4:in `<top (required)>'
	6: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:77:in `run'
	5: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/app.rb:123:in `display_time_entries'
	4: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:10:in `format_entries'
	3: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:10:in `map'
	2: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:12:in `block in format_entries'
	1: from /Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:20:in `time_entry_text'
/Users/gkendal/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/dovico-1.4.0/lib/dovico/model/time_entry_formatter.rb:36:in `project_task': undefined method `tasks' for nil:NilClass (NoMethodError)
```

**Comment by chtitux** [comment]
Well, maybe it's still does not work because I missed something but it's hard for me to know because I can't reproduce pagination.

Could you test from a clean code base, so we're sure you're using the branch:

~~~bash
$ git clone -b pagination https://github.com/chtitux/dovico.git
$ cd dovico/
$ bundle exec bin/dovico --tasks
$ dovico --show --start 2019-10-01 --end 2019-10-07
~~~

and tell me if it's working or not.
If it's not, could you send, if possible me to chtitux (at) gmail (dot) com the result of the 2 commands please?
I'll try to see where is the bug.

**Pull request merged**